### ISSUE TITLE
How to cluster words into semantic entities, when performing information extraction?

### ISSUE THREAD

**Issue by FrancoisNoyez** [NielsRogge/Transformers-Tutorials#214]
Hi everyone,

I've got a question regarding information extraction from forms. Not sure if it's the right location to use to ask it. Don't hesitate to tell me if you think this post should be made elsewhere instead.

I've been searching the web about this for some time now, but I have not found a satisfactory answer yet.
My objective is to automatically extract the content of a form as a list of (QUESTION; ANSWER) pairs.

There are tutos and notebook dedicated to showing how to perform information extraction by using LayoutLMForTokenClassification (ex: [here](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForTokenClassification_on_FUNSD.ipynb); or [here](https://www.philschmid.de/fine-tuning-layoutlm)).

However, with this, all we achieve is the labeling of words.
To actually achieve my desired result, we need to do two more things:

A] First, we need to group together words into semantic entities; i.e. to group together the words which make up the name of a field / make up a QUESTION; or the words which constitute an ANSWER (for instance: first the word with the value and then the word for the unit, e.g. "3.5 kg").

B] Then we need to determine / extract the relations between those entities. For instance, we need to find the relation linking one semantic entity which is a QUESTION, with one which is the associated ANSWER (which may also not exist).

There already exist some resources dedicated to performing the RelationExtraction part.
Ex: by the people of Microsoft working on unilm, who implemented the LayoutLMv2ForRelationExtraction model (one tuto notebook is available [here](https://colab.research.google.com/drive/16wqA3oTUf7yzUKsSSZxiMf1443_ZO3wC?usp=sharing#scrollTo=iKEVpigK6Uq4)). In fact, one PR had been opened on the "transformers" repo in order add this model class. The latest version of the PR is [here](https://github.com/huggingface/transformers/pull/18943).

However, in order to work, this model assumes that we already know the semantic entities, hence the need for the A] step.
But then, how do we achieve this step? That's what I'm struggling with.
The output of the use of the LayouLMTokenClassification will look something like this:
![display_words_boxes](https://user-images.githubusercontent.com/47528215/202765832-1923445a-fdee-45a2-9434-d501c4cfc87b.png)

However, to use the RelationExtraction model, we need something like this instead.
![display_entities_boxes](https://user-images.githubusercontent.com/47528215/202765857-d2e604a0-c25a-41b1-9cd2-25bb62376ad5.png)

Sure, we have this info for labelled data, but not at inference time, for a whole new document.
So, how do we do that for new documents? The order in which the words is output by the OCR may not be consistent with the order in which we actually need to consider the words, if we just rely on the label values to perform the decoding / the build of the semantic entities.

We can model the problem in various ways: notably, this can be though up as finding the edges between the nodes of a directed graph, where the indegree and outdegree values of each vertex are at most 1, and there is no cycle.

I have crafted a simple algo based notably on computing the distance between the bounding box of the words, to account for the info contained in the spatial locations of the words relative to one another, but it's far from perfect, and there are cases when it will fail to produce the correct result.

Seems to me like it's the job the LayouLM -based model in the first place to consider both the semantic information and the spatial information of the words, in order to perform the proper labeling within the LayoutLMForTokenClassification model (notably, to distinguish the "B-" labels from the "I-" label, if we refer to the FUNSD dataset), so it's a shame that this info ("to which semantic entity belongs each token / each word") is not output by it.

Does anyone have any idea regarding how to carry out the step A]?
Notably, do we need to craft a dedicated model, such as LayoutLMForTokenClassification and LayoutLMv2ForRelationExtraction, in order to achieve this? Or is it possible to somehow re-use / upgrade the LayoutLMForTokenClassification model, in order to produce outputs which allows to carry out the semantic entities construction task?

**Comment by JaMe76** [comment]
Hi,

I think you have a valid point here. You might get away with `LayoutLMForTokenClassification` trained on a mix of `B,I,O` + `question/answer/header/other` combination provided you have an easy structured form, where the OCR output might be almost consistent with the reading order of the entities.
However, the approach will fall apart the more entities are spread over multi lines and key/pairs are scattered over several columns (e.g. your example).
     
Unfortunately, I do not have an easy solution either. I’ve been thinking about training an object detector for detecting entities. With respect to FUNSD you have bounding boxes on entity level, so training data is available, but of course, this will also add another model to your stack. On the other hand, you will certainly be able to build something robust. 

Nevertheless, I have a real interest to a working solution, so if you’re interested feel free to reach out. I am working on a package that is dedicated to provide some end-to-end solutions to these types of problems: 

https://github.com/deepdoctection/deepdoctection .
 
There are some options to use `LayoutLMv1` for the standard problems, but the key-value extraction is on my agenda right now and I would be happy to discuss some ideas.

**Comment by FrancoisNoyez** [comment]
Hi Jame76,

Thank you for your answer.

Yes, for my first use cases I think the algo that I created will be ok most of the time; but I still wish I had a more reliable solution right away.

Adding another model to the stack is ok for now: I first want to have a proper, working pipeline to feed the inference data into, I'll worry about making it time and resource efficient later. There are already two models in the stack, one for token classification and one for relation extraction, so no big deal regarding inserting a third one between those two, if necessary.
Also, yes, I guess that if we were able to first detect the semantic entities using object detection, things would be easier. However, object detector will typically use only visual information; whereas, in order to properly detect the entities (notably, to properly separate a question from its answer in some cases), it would be useful or even necessary to consider the semantic information as well. Combining the two sources of info was the point of LayoutLM architecture, so I guess we then need to craft a model similarly to how the LayoutLMForTokenClassification model was made, by adding a custom layer on top of LayoutLM, but this time asking the model to output also the info needed to link together words into the proper semantic entities. To do that, I guess we could draw inspiration from the RelationExtraction model, since its point is also to find relationships between text elements, and somehow "link them together". We can't reuse it as it is (at least, as it was at the time when the tuto notebook was created, it's the only one that I have experimented with up until now), because the only kind of relationship that it accepts to deal with for now is the one existing between a QUESTION and an ANSWER (so, different type of entities, and can have only two elements / entities in the relationship). I need to look further into it to actually understand the layer that was added on top of LayoutLM, and how to modify it to achieve the desired result.

Got it, thank you for sharing about your deepdoctection project. I find that the model available for demonstration on the huggingface space allows to get impressive results. Notably regarding table extraction; that could already be useful for some use cases that I intended to tackle later in the future.
